core/src/main/scala/org/apache/spark/rdd/LocalRDDCheckpointData.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rdd
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.{Logging, SparkEnv, SparkException, TaskContext}
+import org.apache.spark.storage.{RDDBlockId, StorageLevel}
+
+/**
+ * An implementation of checkpointing implemented on top of Spark's caching layer.
+ *
+ * Local checkpointing trades off fault tolerance for performance by skipping the expensive
+ * step of saving the RDD data to a reliable and fault-tolerant storage. Instead, the data
+ * is written to the local, ephemeral block storage that lives in each executor. This is useful
+ * for use cases where RDDs build up long lineages that need to be truncated often (e.g. GraphX).
+ */
+private[spark] class LocalRDDCheckpointData[T: ClassTag](@transient rdd: RDD[T])
+  extends RDDCheckpointData[T](rdd) with Logging {
+
+  /**
+   * Ensure the RDD is fully cached so the partitions can be recovered later.
+   */
+  protected override def doCheckpoint(): CheckpointRDD[T] = {
+    val level = rdd.getStorageLevel
+
+    // Assume storage level uses disk; otherwise memory eviction may cause data loss
+    assume(level.useDisk, s"Storage level $level is not appropriate for local checkpointing")
+
+    // Not all RDD actions compute all partitions of the RDD (e.g. take)
+    // We must compute and cache any missing partitions for correctness reasons
+    // TODO: avoid running another job here (SPARK-8582)
+    rdd.count()
```

Comment:
I mentioned in the earlier comment thread, you may have missed it. Recommenting it here.

This `rdd.count` isnt great either. Even when its cached, it may be cached on disk, or serialized in memory. In which case running a count may be costly and time consuming, and pretty much defeats the purpose of making this a cheap checkpointing. Also, in the majority of the cases, this will be fully cached, in which case running this job is superfluous. The right thing to do (which aint too hard) is to find out which partitions are missing and only run those partitions.
```
val missingPartitionIds = rdd.partition.filter { p => 
   !blockManagerMaster.contains(RDDBlockId(rdd.id, p.index)) 
}.map { _.index }

rdd.sparkContext.runJob(
  rdd, 
  (tc: TaskContext, iterator: Iterator[T]) => Utils.getIteratorSize(iterator)  // same as count()
  missintPartitionIds
)
```